core/drivers/ls_dspi.c

```diff
@@ -0,0 +1,450 @@
+			dspi_rx(data);
+		} else if (rdata) {
+			dspi_tx(data, ctrl, DSPI_IDLE_DATA);
+			*spi_rd++ = dspi_rx(data);
+		}
+		num_pkts = num_pkts - 1;
+	}
+
+	return SPI_OK;
+}
+
+static enum spi_result nxp_dspi_txrx16(struct spi_chip *chip, uint16_t *wdata,
+				       uint16_t *rdata, size_t num_pkts)
+{
+	uint32_t ctrl = 0;
+	uint16_t *spi_rd = NULL, *spi_wr = NULL;
+
+	spi_wr = wdata;
+	spi_rd = rdata;
+
+	struct nxp_dspi_data *data = container_of(chip, struct nxp_dspi_data,
+						  chip);
+	unsigned int cs = data->slave_cs;
+
+	/*
+	 * Assert PCSn signals between transfers
+	 * select which CTAR register and slave to be used for TX
+	 * CTAS selects which CTAR to be used, here we are using CTAR0
+	 * PCS (peripheral chip select) is selecting the slave.
+	 */
+	ctrl = DSPI_TFR_CTAS(data->ctar_sel) | DSPI_TFR_PCS(cs);
+	if (data->slave_mode & SPI_CONT)
+		ctrl |= DSPI_TFR_CONT;
+
+	if (data->slave_data_size_bits != 16) {
+		EMSG("data_size_bits should be 16, not %u",
+		     data->slave_data_size_bits);
+		return SPI_ERR_CFG;
+	}
+
+	while (num_pkts) {
+		if (wdata && rdata) {
+			dspi_tx(data, ctrl, *spi_wr++);
+			*spi_rd++ = dspi_rx(data);
+		} else if (wdata) {
+			dspi_tx(data, ctrl, *spi_wr++);
+			dspi_rx(data);
+		} else if (rdata) {
+			dspi_tx(data, ctrl, DSPI_IDLE_DATA);
+			*spi_rd++ = dspi_rx(data);
+		}
+		num_pkts = num_pkts - 1;
+	}
+
+	return SPI_OK;
+}
+
+static void nxp_dspi_start(struct spi_chip *chip)
+{
+	struct nxp_dspi_data *data = container_of(chip, struct nxp_dspi_data,
+						  chip);
+
+	DMSG("Start DSPI Module");
+	io_clrbits32(data->base + DSPI_MCR, DSPI_MCR_HALT);
+}
+
+static void nxp_dspi_end(struct spi_chip *chip)
+{
+	struct nxp_dspi_data *data = container_of(chip, struct nxp_dspi_data,
+						  chip);
+
+	/* De-assert PCSn if in CONT mode */
+	if (data->slave_mode & SPI_CONT) {
+		unsigned int cs = data->slave_cs;
+		unsigned int ctrl = DSPI_TFR_CTAS(data->ctar_sel) |
+				    DSPI_TFR_PCS(cs);
+
+		/* Dummy read to deassert */
+		dspi_tx(data, ctrl, DSPI_IDLE_DATA);
+		dspi_rx(data);
+	}
+
+	DMSG("Stop DSPI Module");
+	io_setbits32(data->base + DSPI_MCR, DSPI_MCR_HALT);
+}
+
+void dspi_flush_fifo(struct nxp_dspi_data *dspi_data)
+{
+	unsigned int mcr_val = 0;
+
+	mcr_val = io_read32(dspi_data->base + DSPI_MCR);
+
+	/* flush RX and TX FIFO */
+	mcr_val |= (DSPI_MCR_CTXF | DSPI_MCR_CRXF);
+
+	io_write32(dspi_data->base + DSPI_MCR, mcr_val);
+}
+
+static void dspi_set_cs_active_state(struct nxp_dspi_data *dspi_data,
+				     unsigned int cs, unsigned int state)
+{
+	DMSG("Set CS active state cs=%d state=%d", cs, state);
+
+	if (state & SPI_CS_HIGH)
+		/* CSx inactive state is low */
+		io_clrbits32(dspi_data->base + DSPI_MCR, DSPI_MCR_PCSIS(cs));
+	else
+		/* CSx inactive state is high */
+		io_setbits32(dspi_data->base + DSPI_MCR, DSPI_MCR_PCSIS(cs));
+}
+
+static void dspi_set_transfer_state(struct nxp_dspi_data *dspi_data,
+				    unsigned int state)
+{
+	DMSG("Set transfer state=%d bits=%d", state,
```

Comment:
Move below declaration(s)

Reply:
Comment not addressed

Reply:
> Comment not addressed
Apologies for oversight. It is addressed now.